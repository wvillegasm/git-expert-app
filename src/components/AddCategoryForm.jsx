import { useState } from "react";

export const AddCategoryForm = ({ onAddCategory }) => {
  const [gifCategory, setGifCategory] = useState("");

  const handleOnChange = (e) => {
    setGifCategory(e.target.value);
  };

  const handleOnSubmit = (e) => {
    e.preventDefault();

    const gifCategoryFormatted = gifCategory.trim().toUpperCase();

    if (gifCategoryFormatted.length < 2) {
      return;
    }

    onAddCategory(gifCategoryFormatted, setGifCategory);

    // setGifCategory("");
  };

  return (
    <form className="space-bottom" onSubmit={handleOnSubmit}>
      <input
        type="text"
        placeholder="Search gifs"
        onChange={handleOnChange}
        value={gifCategory}
      />
    </form>
  );
};
