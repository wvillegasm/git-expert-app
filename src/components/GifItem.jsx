export const GifItem = ({ title, url }) => {
  return (
    <div className='gif-item'>
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};
