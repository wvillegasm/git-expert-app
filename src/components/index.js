export * from './AddCategoryForm';
export * from './GifGrid';
export * from './GifItem';
