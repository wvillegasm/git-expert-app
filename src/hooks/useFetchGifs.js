import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
  const [images, setImages] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    let isMounted = true;

    const fetchImages = async () => {
      setIsLoading(true);
      try {
        const images = await getGifs(category);
        if (isMounted) {
          setImages(images);
        }
      } catch (error) {
        console.error('Error fetching gifs:', error);
        if (isMounted) {
          setImages([]);
        }
      } finally {
        if (isMounted) {
          setIsLoading(false);
        }
      }
    };

    fetchImages();

    return () => {
      isMounted = false;
    };
  }, [category]);

  return {
    images,
    isLoading,
  };
};
