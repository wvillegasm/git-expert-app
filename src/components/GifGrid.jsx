import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem } from './';

import '../styles.css';

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>

      {isLoading && <p>Loading...</p>}

      <div className='gif-grid'>
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
