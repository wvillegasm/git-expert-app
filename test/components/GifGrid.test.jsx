import { render, screen } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import { GifGrid } from "../../src/components/GifGrid";
import { useFetchGifs } from "../../src/hooks/useFetchGifs";

// Mock the custom hook
vi.mock("../../src/hooks/useFetchGifs");

describe("GifGrid Component", () => {
  const category = "Dragon Ball";

  test("should show loading message initially", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true,
    });

    render(<GifGrid category={category} />);
    expect(screen.getByText("Loading...")).toBeTruthy();
    expect(screen.getByText(category)).toBeTruthy();
  });

  test("should show items when images are loaded", () => {
    const gifs = [
      {
        id: "ABC123",
        title: "Goku",
        url: "https://localhost/goku.gif",
      },
      {
        id: "DEF456",
        title: "Vegeta",
        url: "https://localhost/vegeta.gif",
      },
    ];

    useFetchGifs.mockReturnValue({
      images: gifs,
      isLoading: false,
    });

    render(<GifGrid category={category} />);

    expect(screen.getByText(category)).toBeTruthy();
    expect(screen.getAllByRole("img").length).toBe(2);
    expect(screen.queryByText("Loading...")).toBeNull();
  });

  test("should call useFetchGifs with correct category", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true,
    });

    render(<GifGrid category={category} />);
    expect(useFetchGifs).toHaveBeenCalledWith(category);
  });

  test("should match snapshot", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true,
    });

    const { container } = render(<GifGrid category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("should have gif-grid class on container", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: false,
    });

    render(<GifGrid category={category} />);
    expect(screen.getByRole("heading", { level: 3 }).textContent).toBe(
      category
    );
    expect(document.querySelector(".gif-grid")).toBeTruthy();
  });
});
