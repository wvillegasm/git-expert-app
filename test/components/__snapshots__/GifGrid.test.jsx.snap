// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GifGrid Component > should match snapshot 1`] = `
<div>
  <h3>
    Dragon Ball
  </h3>
  <p>
    Loading...
  </p>
  <div
    class="gif-grid"
  />
</div>
`;
